keps/sig-api-machinery/4355-coordinated-leader-election/README.md

```diff
@@ -0,0 +1,1319 @@
+
+### Enabling on a component
+
+Components with a `--leader-elect-resource-lock` flag (kube-controller-manager,
+ kube-scheduler) will accept `coordinatedleases` as a resource lock type.
+
+### Migrations
+
+So long as the API server is running a coordinated election controller, it is
+safe to directly migrate a component from Lease Based Leader Election to
+Coordinated Leader Election (or vis-versa).
+
+During the upgrade, a mix of components will be running both election
+approaches. When the leader lease expires, there are a couple possibilities:
+
+- A controller instance using Lease Based Leader Election claims the leader
+  lease
+- The coordinated election controller picks a leader, from the components that
+  have written identity leases, and claims the lease on the leader's behalf
+
+Both possibilities have acceptable outcomes during the migration-- a component
+is elected leader, and once elected, remains leader so long as it keeps the
+lease renewed. The elected leader might not be the leader that Coordinated
+Leader Election would pick, but this is no worse than how leader election works
+before the upgrade, and once the upgrade is complete, Coordinated Leader
+Election works as intended.
+
+There is one thing that could make migrations slightly cleaner: If Coordinated
+Leader Election adds a `coordination.k8s.io/elected-by:
+leader-election-controller` annotation to any leases that it claims. It can also
+check for this annotation and only mark leases as "end-of-term" if that
+annotation is present. Lease Based Leader Election would ignore "end-of-term"
+annotations anyway, so this isn't strictly needed, but it would reduce writes
+from the coordinated election controller to leases that were claimed by
+component instances not using Coordinated Leader Election
+
+### API
+
+The lease lock API will be extended with a new field for election preference, denoted as an enum for strategies for Coordinated Leader Election.
+
+```go
+
+type CoordinatedLeaseStrategy string
+
+// CoordinatedLeaseStrategy defines the strategy for picking the leader for coordinated leader election.
+const (
+  OldestCompatibilityVersionStrategy CoordinatedStrategy = "OldestCompatibilityVersionStrategy"
+  NewestCompatibilityVersionStrategy CoordinatedStrategy = "NewestCompatibilityVersionStrategy"
+)
+
+type LeaseSpec struct {
+	// Strategy indicates the strategy for picking the leader for coordinated leader election
+	Strategy CoordinatedLeaseStrategy `json:"strategy,omitempty" protobuf:"string,6,opt,name=strategy"`
+
+	// EndofTerm signals to a lease holder that the lease should not be
+	// renewed because a better candidate is available.
+	EndOfTerm bool `json:"endofterm,omitempty" protobuf:"boolean,7,opt,name=endofterm"`
```

Comment:
There is no acknowledge of this, neither during the transition to coordinated leader election, but also in general. Am I right that we are introducing a race here with a possible time where two leaders might be active?

In the old `Lease` world, a controller could avoid the race by shutting down/pause its work before the lease expires while renewal fails. In the new world, eventually the `endOfTerm` happens at arbitrary times and the leader might not even see it and continue its work.

Hence, I feel like we need acknowledgement, i.e. the current leader must step down by unsetting `HolderIdentity`.

Reply:
Oh, ic. Missed that. I thought it is an ask to give up the lease earlier.

Reply:
It is an ask, but a controller can choose when to give it up. Controllers would need to shut down its work before giving up though so we're opting to have it act as a "do not renew" marker for now and it could be optimized in the future to release the lease sooner.